File: 01-Languages/CSharp/22-Mini-Projects/TaskTracker/TaskItem.cs
namespace TaskTracker;

// Priority/Status as enums (not raw strings) so an invalid value is a
// compile-time error at every call site, and the SQLite layer only ever
// has to translate a small closed set of values, not validate free text.
public enum Priority { Low, Medium, High }

public enum Status { Pending, Done }

// A record rather than a class: task rows are read back from SQLite as
// fresh snapshots on every query, never mutated in place, so value
// equality (two tasks with identical fields are "the same task") is the
// right default and comes for free with a record.
public record TaskItem(int Id, string Title, Priority Priority, Status Status, DateTime CreatedAt) {
    public override string ToString() {
        var statusMark = Status == Status.Done ? "[x]" : "[ ]";
        return $"{statusMark} #{Id,-3} {Title,-30} priority={Priority,-6} created={CreatedAt:yyyy-MM-dd}";
    }
}

File: 01-Languages/Kotlin/22-Mini-Projects/TaskTracker/src/TaskItem.kt
// Enums instead of free-text strings for priority/status -- an invalid priority is a compile
// error here, not a runtime-discovered typo in a database row (the exact problem free-text
// columns invite).
enum class Priority { LOW, MEDIUM, HIGH }
enum class Status { PENDING, DONE }

// A data class rather than a plain class specifically so tests get real structural equality
// (Exercise 02's finding, reused here) -- comparing two TaskItems in an assertion compares their
// field values, not their object identity, which is what a test actually wants to check.
data class TaskItem(
    val id: Long,
    val title: String,
    val priority: Priority,
    val status: Status,
    val createdAt: String, // stored as SQLite's own ISO-8601 default (see TaskRepository), kept as a
                            // plain String rather than a proper date type to avoid pulling in a third
                            // dependency (java.time formatting) just for this mini-project's display needs
)

data class TaskStats(val pending: Int, val done: Int, val total: Int)
